### ISSUE TITLE
Projections are not durable

### ISSUE THREAD

**Issue by maxekman** [looplab/eventhorizon#32]
Imagine your projection is down and several commands reach to the command side and result into some events persisted into event store. Once the projection is up again, new events will not be collected.

**Comment by maxekman** [comment]
@dtravin FYI, added this as an issue from our mail conversation.

**Comment by bweston92** [comment]
When a projection goes down it should either know the position of the last event in the stream.

Or should just clear any previous state that was made by projection and run it again then mark it self as ready before accepting requests etc.

**Comment by nmkarvekar** [comment]
@maxekman @bweston92   Is it applicable even if projections and event store are part of same microservice? would like to understand, even if both are part of same microservice, is it possible that only projector gets down?  or it it the case when mogodb projections and event store are in different mongo cluster?  Also as I understand, writing the event to event store and projections is atomic operation.

**Comment by bweston92** [comment]
Writing to an event store and projections is and should never be atomic.

You ideally would have a write service and a query service that way you can scale each independently. You can then also have many different queryable outputs from the events.

**Comment by nmkarvekar** [comment]
thanks @bweston92 . Would like to know the use case when only particular projections  gets down? Is it possible even if they are part of same microservice?

**Comment by maxekman** [comment]
Yes, if the underlying storage is not responding for example. Remember that the projections could use a different storage than the event store, and as @bweston92 noted there could even be multiple projections for the same event stream (MongoDB and BigQuery could be one such example).

**Comment by Encounter27** [comment]
@maxekman @bweston92 we are evaluating technology stack for one of our projects and I found that event horizon can solve our need of CQRS/ES in Golang 😊.

What I understood about the current flow of Eventhorizon: -
![image](https://user-images.githubusercontent.com/8401256/59379222-bb060480-8d73-11e9-8329-e4bfef9d392c.png)

So, I was thinking about this since last couple of days, to solve the bellow problems:
1.	Store-Event and projector should be independent, so write side and read side can be scale separately.
2.	Irrespective of read side projectors (There can be multiple read side projectors) the write side should be 100% available.
3.	Indeed, we should be able to publish events to some message queue from one of the read side projectors.

So, I came up with the below design: -

![image](https://user-images.githubusercontent.com/8401256/59379263-cfe29800-8d73-11e9-95d3-150d7a283dcb.png)

So, in the above diagram, Two aggregate is there A1, A2. On A1 [A1:E1, A1:E2, A1:E3] events applied and on A2 [A2:E1, A2:E2] events applied.
1.	The event store is the same event store present in event horizon. 

![image](https://user-images.githubusercontent.com/8401256/59380456-c73f9100-8d76-11e9-9101-6173a4ade506.png)


2.	Event outbox will have each event (having the ISO timestamp) occurred in the system. These events are sorted by timestamp.

![image](https://user-images.githubusercontent.com/8401256/59379694-e3dac980-8d74-11e9-9e73-bee0dbda5133.png)


3.	Writing to event store and outbox will be atomic.
4. Time offset collection will hold the timestamp for each projector till the projector processed the events.
![image](https://user-images.githubusercontent.com/8401256/59379917-85621b00-8d75-11e9-89de-56bbcb39ff95.png)


Now, say we need 3 projectors:
1.	Each projector is independent go-routine.
2.	Say projector1 routine prepare read side view in //DB1.
And projector2 routine prepare read side view in //DB2.
Projector3 routine publish event to some msg queue say Kafka.

Lifecycle/Responsibility of each projection routine:
1.	When the application will start all the projection routine will also start and keep running until the pod stops or pod crashes.
2.	Now each projector will check the Time offset collection. if the projector entry is not found it will start read from the first event and update the projector entry and with the timestamp till it processed.
3.	Also, I added a string field “projectionDone” to each event present in outbox. So, the length of the field will be equal to number of read side projector.
For total three projector, for each event of outbox the default value of “projectionDone” field will be “FFF”. 0-index for projector1, 1-index char for projector2 and so on.
4.	So, a projector when it will read an event it will mark the field with ‘I’ (In process). Then the projector will write the event to read side DB/Kafka and if succeed, it will mark the field to ‘T’ (Processed) and also update the Time offset.
Example: projector1
“FFF” -> “IFF” -> “TFF” 
Also this char position will help to synchronize the same projector present in multiple pods (See the diagram).
Here the above step 2, 3, 4 will be atomic for one projector.

Benefits:
1.	If we reset the time in Time Offset collection say to initial state, we will be able to replay all the projections.
2.	Now write side and read side can be scale separately.

Concern:
1.	Read side projection delay. May be this eventual delay is ok as per CAP theorem.
2.	If same projector is running parallelly in multiple pods, there will be some contention, and these doesn’t make any sense as one read side projection should process one event only once.
For this I am thinking below sol:
```Run the projector completely in separate and single pod. So basically, for write there can be multiple pod and for projection there will be only one pod```

Please review and share the thought on this. I am not sure how much I could express in words :) , may be for more clarification we can have a short video call. 
If the idea sounds good to you all, I will create a PR.

Waiting for some feedback…

Thanks,
Pradip

**Comment by maxekman** [comment]
@encounter27 Can you please post this as a new issue, this is a bug report. Thanks.
